### PR TITLE
Compute Multiplicative Inverse over Secp256k1 Base Field

### PR DESCRIPTION
**Pull request by itzmeanjan**
This routine is required for implementing ECDSA signature verification over secp256k1 curve. To be more specific, we need to be able to convert a secp256k1 curve point from its projective coordinate form to affine coordinate system.

> **Note**
> See how that's done, [here](https://github.com/itzmeanjan/secp256k1/blob/37b339db3e03d24c2977399eb8896ef515ebb09b/point/point.py#L55-L65)

Routine | VM Cycles
:-- | --:
Mul over secp256k1 base field | 1498
MulInv over secp256k1 base field | 782,413

---

Expected review order: #481 -> #482 -> `this`

### CODE REVIEW

**Comment by itzmeanjan** [comment]
Resolved merge conflicts at https://github.com/maticnetwork/miden/pull/487/commits/c138a270a588228e0d46edc9f1c12f730695641b

**Review by bobbinth — CHANGES_REQUESTED** [review]
Looks great! Thank you! I left one question inline, but also had a couple of general comments:

First, I think we should probably change the namespace structure a bit. Specifically, instead of having 3 `secp256k1`-related modules in the `math` namespace, we should probably put all these modules into the common namespace. For example, something like this:

- `std::math::secp256k1::base_field`
- `std::math::secp256k1::scalar_field`
- `std::math::secp256k1::ec` or maybe `std::math::secp256k1::curve`

Second, and this is more comment for the future: I'd like to rely on external libraries for testing as much as possible. So, if there is a well-tested library out there which would allow us to do base field, scalar field etc. arithmetic - would be great to use it. Maybe create an issue for this?

File: miden/tests/integration/stdlib/math/secp256k1_base_field.rs [hunk 1]
```diff
@@ -0,0 +1,603 @@
+use super::{build_test, Felt};
+use std::cmp::PartialEq;
+use std::ops::{Add, Mul, Neg, Sub};
+
+/// Secp256k1 base field element, kept in Montgomery form
+#[derive(Copy, Clone, Debug)]
+struct BaseField {
+    limbs: [u32; 8],
+}
```
[the remark this answers is not in the document]
Comment (itzmeanjan) [reply miden/tests/integration/stdlib/math/secp256k1_base_field.rs:9]: Sure we'll do that, it's noted down https://github.com/0xPolygonMiden/crypto/issues/19

File: miden/tests/integration/stdlib/math/secp256k1_base_field.rs [hunk 2]
```diff
@@ -0,0 +1,603 @@
...
+    begin
+        exec.secp256k1_base_field::add
+    end";
+
+    let source_sub = "
+    use.std::math::secp256k1_base_field
+
+    begin
+        exec.secp256k1_base_field::sub
+    end";
```
[the remark this answers is not in the document]
Comment (itzmeanjan) [reply miden/tests/integration/stdlib/math/secp256k1_base_field.rs:508]:
```
random a ∈ secp256k1 base field
random b ∈ secp256k1 base field

c = a + b
d = c - a

assert d == b
```

With this execution flow it checks correctness of both Rust and Miden assembly implementation of sepc256k1 scalar field arithmetic, which could have been written differently. 

Note, these commits come from some other PR, where it's already being improved. So I'd rather not touch it.
See  https://github.com/maticnetwork/miden/commit/5b68ec5b2ec5f8b245e800857ff9dce1fb93e806 where I touch them.

**Comment by itzmeanjan** [comment]
> Looks great! Thank you! I left one question inline, but also had a couple of general comments:
> 
> First, I think we should probably change the namespace structure a bit. Specifically, instead of having 3 `secp256k1`-related modules in the `math` namespace, we should probably put all these modules into the common namespace. For example, something like this:
> 
>     * `std::math::secp256k1::base_field`
> 
>     * `std::math::secp256k1::scalar_field`
> 
>     * `std::math::secp256k1::ec` or maybe `std::math::secp256k1::curve`
> 
> 
> Second, and this is more comment for the future: I'd like to rely on external libraries for testing as much as possible. So, if there is a well-tested library out there which would allow us to do base field, scalar field etc. arithmetic - would be great to use it. Maybe create an issue for this?

Good point, I've renamed them like

- Base field arithmetic : `std::math::secp256k1::base_field`
- Scalar field arithmetic : `std::math::secp256k1::scalar_field`
- Curve arithmetic : `std::math::secp256k1::group`

See 0ce38e6a953282e4950031f0ee094a38f737e0bc

**Comment by itzmeanjan** [comment]
Speaking of extracting out cryptographic constructions from tests and putting them in crypto crate, see https://github.com/0xPolygonMiden/crypto/issues/19

**Review by bobbinth — APPROVED** [review]
All looks good! Thank you!

**Pull request merged**